### ISSUE TITLE
OSCraw does not receive arrays via OSC

### ISSUE THREAD

**Issue by Oeyvind** [csound/csound#1641]
If I send an array with OSCsend, OSCraw will produce an array showing the message type to be "b" (bundle?), and then just empty strings for the rest of the array (nothing is copied I guess).

```
<CsoundSynthesizer>
<CsOptions>
-odac
</CsOptions>
<CsInstruments>

sr = 48000
ksmps = 128
0dbfs = 1
nchnls = 2

instr    1
  kval line 440, p3, 880
  kSend[] init 2
  kSend[0] = kval
  ksend metro 10
  OSCsend ksend+1, "127.0.0.1", 9999, "/foo", "A", kSend

  Smess[] init 8
  nxtmsg:
    Smess[], klen OSCraw 9999
    if klen > 0 then
      printarray Smess
      kgoto nxtmsg
    endif
endin

</CsInstruments>
<CsScore>
i1 0 10
</CsScore>
</CsoundSynthesizer>

```

**Comment by vlazzarini** [comment]
Looking at the code, it seems to try to get the first number in the array as its size

```
          } else if (c == 'b') {
            len = *((uint32_t *) buf);
            byteswap((char*)&len,4);
            len = ceil((len)/4.)*4;
            if (len > str[n].size) {
              str[n].data = csound->ReAlloc(csound, str[n].data, len+1);
              str[n].size  = len+1;
            }
            strNcpy(str[n].data, buf, len+1);
            //str[n].data[len] = '\0';
            buf += len;
          }
```

Not sure whether this is the right thing, but we need to know how long the array is, and I don't know if there is another way.

**Comment by kunstmusik** [comment]
@vlazzarini @Oeyvind : Checking to see if this is still an issue; if so I will schedule for 7.0.0.

**Comment by Oeyvind** [comment]
@kunstmusik @vlazzarini : Yes I think it is still an issue. I haven't trioed anythng more with this, as I understood from @vlazzarini answer that sending arrays would not work with the current version of the opcode. It is possible that I misunderstood the reply(?). I haven't looked into the details of sending arrays over OSC in other languages. If it is possible in other languages, it would be good if we can adpot a similar strategy/syntax.
* did just a quick search: ..., and it seems python-osc does not support sending arrays (https://stackoverflow.com/questions/42427812/python-osc-message-sending-to-supercollider), but that supercollider might support it in its communication between SC client and SC server
*  ... and here (https://forum.openframeworks.cc/t/get-array-data-from-max-msp-with-osc-protocol/17975) it seems like Max/Msp can send arrays over OSC but that Openframeworks does not have a ready function to receive it. BUT ALSO that they suggest using m.getNumArgs() in openframeworks to get the length of the array and thus be able to process it. Perhaps this is the way @vlazzarini suggested we should do it(?)

**Comment by vlazzarini** [comment]
as I said, you can try sending a numeric with its first number being the length of the array. That seems to be the way the code is conceived.

**Comment by vlazzarini** [comment]
I checked and "b" is boolean. So my answer was confusing, "#bundle" is a type that sends its size first, then data, but I don't think it is relevant here.

OSCraw supports f, i, b, s. No other type. 

I looked to see, and arrays are supported by OSCsend but not by OSClisten (even though the manual says it does accept it).
@jpff do you know why?

The array is sent as:

dimensions
size
data 

The only problem I see is that we don't know whether the data is floats or doubles. So it only works between two Csound instances of the same float version.

Also, it is a proprietary type, defined for Csound arrays only.

I see also the type 'D', which is 

size
data

All of this in OSCsend.

**Comment by vlazzarini** [comment]
Addressed by https://github.com/csound/csound/pull/2186
